### PR TITLE
[WIP] Restrict who is able to join a team

### PR DESCRIPTION
**Pull request by Ppjet6**
Restrictions are applied per domain. Use team.AllowedDomains if present,
otherwise use the global RestrictCreationToDomains.

#### Summary

As stated in #6943, this will allow admins or team owners to manage who has access to teams.

This first commit deals with the logic part. I still need to write tests, and I would also like to add a bit of UI in the team settings.

#### Ticket Link
#6943 

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by jasonblais** [comment]
Thanks @Ppjet6 for this. Sorry I've been busy with getting 4.0.0 out but I'll review this soon, in a day or two.

Let's also discuss more about #6941 and #6943 at the same time.

Really appreciate it.

**Comment by jasonblais** [comment]
Thanks again @Ppjet6 for this.

Some questions:
1 - For the following case
 - server: mattermost.com, mattermost.org
 - team: mattermost.com, mattermost.ca

Accounts with mattermost.com and mattermost.org domains will be able to join the server, but only mattermost.com domain will be able to join the team. 

Should there be any feedback that mattermost.ca cannot join the team (since they're blocked at the system level)

2 - Are we planning to restrict this both in the APIs and in the UI (ie. team invite link, email invite link don't work + the users that don't have the accepted domain won't show up in "Add Members to Team" dialog)?

3 - Do you feel https://github.com/mattermost/platform/issues/6941 should be implemented in conjunction with this feature? Because otherwise you could have restricted teams, but the people would still be able to start a direct message conversation.

PS: As a reminder, this might be an Enterprise feature as discussed earlier. Just want to make sure it's on the table.

**Comment by Ppjet6** [comment]
> 1 - For the following case
> 
>     server: mattermost.com, mattermost.org
>     team: mattermost.com, mattermost.ca
> 
> Accounts with mattermost.com and mattermost.org domains will be able to join the server, but only mattermost.com domain will be able to join the team.
> 
> Should there be any feedback that mattermost.ca cannot join the team (since they're blocked at the system level)

At the moment I hooked the check on the team setting at "user join team" time, also moving the global setting with it, (and not just at team creation time like it was before). I also added the check on the team setting for invites, alongside the global check still.

The errors are the same, there is not difference between being refused because of the global or the team setting for the user.

This means there is no error client side when a user can't be added to a team because of a restriction (global nor team) I suppose this wasn't implemented because assuming the global setting is not changed, a user cannot get added to a team since they won't be able to join the server in the first place.

If we were to add feedback in the "Team Settings" for the case you pointed out, (i.e., when there is a mismatch), I would limit it to admins and not team creators, as otherwise priviledged information (global setting) might be revealed.

Also we might want feedback indeed for when people get rejected by the team setting.

> 2 - Are we planning to restrict this both in the APIs and in the UI (ie. team invite link, email invite link don't work + the users that don't have the accepted domain won't show up in "Add Members to Team" dialog)?

It seems like a reasonable thing to do, but I would have to look into this. I'll get back to you with more details.

> 3 - Do you feel #6941 should be implemented in conjunction with this feature? Because otherwise you could have restricted teams, but the people would still be able to start a direct message conversation.

Yes these two issues go by pair.

File: model/team.go [hunk 1]
```diff
@@ -275,7 +275,6 @@ func CleanTeamName(s string) string {
 
 func (o *Team) Sanitize() {
 	o.Email = ""
-	o.AllowedDomains = ""
```
Comment (Ppjet6) [inline model/team.go]: Is there a better way to do this?

**Comment by Ppjet6** [comment]
I just pushed changes for the UI part, pretty simple input field, no checks done yet.

Before I found [this line](https://github.com/mattermost/platform/pull/6954#discussion-diff-128074372L278), I had a hard time figuring out why my `allowed_domains` field would come back as empty whenever I submitted. I don't think it's the best solution, this was probably here for security/privacy concerns, and I'd like input on it.

**Comment by jasonblais** [comment]
Thanks for this! I've added the `Setup Test Server` label which should spin up a test server once the build passes.

I'll give this one a try and see how it feels + think of corner cases there might be.

**Comment by jasonblais** [comment]
Hey @Ppjet6! 

I've tested it out with some feedback/questions. I think the approach you've taken is good.

In each case below, the system-wide setting has mattermost.com and test.com as the accepted domains. TeamA has mattermost.com only as the accepted domain.

1 - Suppose TeamA is "public". That is **Team Settings > Allow any user with an account on this server to join this team** is set to `Yes`.

With test.com account, go to the Mattermost Main Menu, then Join Another Team.

Select TeamA from the list.

Observed: The page loads forever with no user feedback

image

Expected: Either
 a) TeamA doesn't appear in the list. (preferred, but not sure if it's technically complicated or if it could affect performance in a Mattermost instance with lots of users and teams)
 b) User is redirected to the same error page as when trying to use the team invite link to join TeamA

2 - In light of our conversation, we'll probably have to keep #6941 separate from this, given it would likely require a major refactoring.

----

The following include additional points from the UX/UI point of view, that I'll spend a bit more time thinking about (feedback/suggestions are welcome):

1) Suppose the test.com receives an invite link and attempts to join the team.

They successfully join the server, but when joining the team, they receive an error page with

```
The invite link was invalid. Please speak with your Administrator to receive an invitation.
```

I'm thinking we'd use something like

```
The email you provided does not belong to an accepted domain. Please contact your administrator or sign up with a different email.
```

2) Users without an accepted domain appear in the "Add Members to Team" list.

image

I'm not sure yet what the best user experience would be. But right now if you try to add a member, it'll fail without any feedback.

One option I was thinking is to have some kind of indicator like `Doesn't belong to an accepted domain` on the user list with the selection disabled. Or not show them at all on the list.

3) I'll think about the UI text on the Team Settings page as well

image

**Comment by jasonblais** [comment]
Hey @Ppjet6, wondering if you had any thoughts on the above?

**Comment by Ppjet6** [comment]
Sorry for the delay!

The two commits have pushed take into account your latest comments.

1. I display an error when trying to join a team where the domain restrictions don't match, (fixing the forever spinning icon issue).
I would have preferred the other way, to remove the team from the list, but that seemed a little bit hearier. Maybe if I get some input/hints about where to look at I can give it a try.

2. Ok.

3. I will try to address this in a future commit.

4. I now filter the users displayed in the "Add Member to Team" list.

**Comment by Ppjet6** [comment]
Is it ok if I rebase the changes now? Apparently tests are not happy because of `TestZZWebSocketTearDown`, and I saw it had been fixed a few days ago.

**Comment by jasonblais** [comment]
Thanks @Ppjet6 

Feel free to rebase. I'll spin up another test server afterwards and test the most recent changes.

**Comment by jasonblais** [comment]
Hey @Ppjet6, I reached out to you via a DM at pre-release.mattermost.com to discuss this contribution.

Really appreciate your efforts on this!

**Comment by jasonblais** [comment]
Chatted with @Ppjet6, 

Until there is a [proper restriction between teams](https://github.com/mattermost/platform/issues/6943), we're hesitant to take the PR for restricting teams by domains. The feature would be misleading as these people could still DM each other in certain cases, and hence add a lot of testing and support work for a feature that doesn't work as intended.

We're currently exploring other alternatives instead, such as single and multi-channel guest accounts https://mattermost.uservoice.com/forums/306457-general/suggestions/9982719-ee-single-channel-guests-for-non-team-members
